fw/cache.c

```diff
@@ -2224,32 +2233,43 @@ cache_req_process_node(TfwHttpReq *req, tfw_http_cache_cb_t action)
 		}
 	}
 
-	resp = tfw_cache_build_resp(req, ce, lifetime, id);
-	/*
-	 * The stream of HTTP/2-request should be closed here since we have
-	 * successfully created the resulting response from cache and will
-	 * send this response to the client (without forwarding request to
-	 * the backend), thus the stream will be finished.
-	 */
-	if (resp && TFW_MSG_H2(req)) {
-		id = tfw_h2_stream_id_close(req, HTTP2_HEADERS,
-					    HTTP2_F_END_STREAM);
-		if (unlikely(!id)) {
-			tfw_http_msg_free((TfwHttpMsg *)resp);
-			tfw_http_conn_msg_free((TfwHttpMsg *)req);
-			goto put;
+	T_DBG("Cache: req->method = %d, req->cache_ctl.flags = 0x%x (is not X-Tempesta-Cache: %d)\n",
+		(int)req->method, req->cache_ctl.flags,
+		(int)((req->cache_ctl.flags & TFW_HTTP_CC_CACHE_PURGE) == 0));
+	if (likely(req->method == TFW_HTTP_METH_GET &&
+		(req->cache_ctl.flags & TFW_HTTP_CC_CACHE_PURGE) == 0)) {
+		resp = tfw_cache_build_resp(req, ce, lifetime, id);
+		/*
+		 * The stream of HTTP/2-request should be closed here since we
+		 * have successfully created the resulting response from cache
+		 * and will send this response to the client (without forwarding
+		 * request to the backend), thus the stream will be finished.
+		 */
+		if (resp && TFW_MSG_H2(req)) {
+			id = tfw_h2_stream_id_close(req, HTTP2_HEADERS,
+						    HTTP2_F_END_STREAM);
+			if (unlikely(!id)) {
+				tfw_http_msg_free((TfwHttpMsg *)resp);
+				tfw_http_conn_msg_free((TfwHttpMsg *)req);
+				goto put;
+			}
 		}
 	}
 out:
-	if (!resp && (req->cache_ctl.flags & TFW_HTTP_CC_OIFCACHED))
-		tfw_http_send_resp(req, 504, "resource not cached");
-	else
+		if (!resp && (req->cache_ctl.flags & TFW_HTTP_CC_OIFCACHED))
+			tfw_http_send_resp(req, 504, "resource not cached");
+		else {
 		/*
 		 * TODO: RFC 7234 4.3.2: Extend preconditional request headers
 		 * if any with values from cached entries to revalidate stored
 		 * stale responses for both: client and Tempesta.
 		 */
-		action((TfwHttpMsg *)req);
+			if (req->method == TFW_HTTP_METH_GET ||
+				unlikely(req->cache_ctl.flags &
+					 TFW_HTTP_CC_CACHE_PURGE)) {
+				action((TfwHttpMsg *)req);
+			}
+		}
```

Comment:
Why to check for GET method? We already has made decision whether to cache the response or not.

The whole logic #1516 is unique for Tempesta and relatively sophisticated, so please leave comments on the steps how `TFW_HTTP_CC_CACHE_PURGE` is processed.

Reply:
I tried to separate processing of PURGE with X-Tempesta-Cache from other cached requests (I wrongly supposed only GET requests were cached).
The difference between processing of common user requests and PURGE with upstream GET is skipping some unnecessary processing like building up the full-body request (we return only HTTP 200 to user).

Reply:
done